govcd/vapp.go

```diff
@@ -91,83 +91,98 @@ func (vapp *VApp) Refresh() error {
-				},
-			},
-		},
-		AllEULAsAccepted: acceptAllEulas,
+	networkConnectionSection := types.NetworkConnectionSection{
+		Info:                          "Network config for sourced item",
+		PrimaryNetworkConnectionIndex: primaryNetworkConnectionIndex,
 	}
 
 	for index, orgVdcNetwork := range orgVdcNetworks {
-		vcomp.SourcedItem.InstantiationParams.NetworkConnectionSection.NetworkConnection = append(vcomp.SourcedItem.InstantiationParams.NetworkConnectionSection.NetworkConnection,
+		networkConnectionSection.NetworkConnection = append(networkConnectionSection.NetworkConnection,
 			&types.NetworkConnection{
 				Network:                 orgVdcNetwork.Name,
 				NetworkConnectionIndex:  index,
 				IsConnected:             true,
 				IPAddressAllocationMode: types.IPAllocationModePool,
 			},
 		)
-		vcomp.SourcedItem.NetworkAssignment = append(vcomp.SourcedItem.NetworkAssignment,
-			&types.NetworkAssignment{
-				InnerNetwork:     orgVdcNetwork.Name,
-				ContainerNetwork: orgVdcNetwork.Name,
-			},
-		)
 	}
 
 	if vappNetworkName != "" {
-		vcomp.SourcedItem.InstantiationParams.NetworkConnectionSection.NetworkConnection = append(vcomp.SourcedItem.InstantiationParams.NetworkConnectionSection.NetworkConnection,
+		networkConnectionSection.NetworkConnection = append(networkConnectionSection.NetworkConnection,
 			&types.NetworkConnection{
 				Network:                 vappNetworkName,
 				NetworkConnectionIndex:  len(orgVdcNetworks),
 				IsConnected:             true,
 				IPAddressAllocationMode: types.IPAllocationModePool,
 			},
 		)
-		vcomp.SourcedItem.NetworkAssignment = append(vcomp.SourcedItem.NetworkAssignment,
-			&types.NetworkAssignment{
-				InnerNetwork:     vappNetworkName,
-				ContainerNetwork: vappNetworkName,
+	}
+
+	return vapp.AddNewVM(name, vappTemplate, &networkConnectionSection, acceptAllEulas)
+}
+
+// AddNewVM adds VM from vApp template with custom NetworkConnectionSection
+func (vapp *VApp) AddNewVM(name string, vappTemplate VAppTemplate, network *types.NetworkConnectionSection, acceptAllEulas bool) (Task, error) {
+
+	if vappTemplate == (VAppTemplate{}) || vappTemplate.VAppTemplate == nil {
+		return Task{}, fmt.Errorf("vApp Template can not be empty")
+	}
+
+	if vappTemplate.VAppTemplate.Children == nil || len(vappTemplate.VAppTemplate.Children.VM) < 1 {
+		return Task{}, fmt.Errorf("vApp Template must have at least one child VM")
+	}
+
+	// Status 8 means The object is resolved and powered off.
+	// https://vdc-repo.vmware.com/vmwb-repository/dcr-public/94b8bd8d-74ff-4fe3-b7a4-41ae31516ed7/1b42f3b5-8b31-4279-8b3f-547f6c7c5aa8/doc/GUID-843BE3AD-5EF6-4442-B864-BCAE44A51867.html
+	if vappTemplate.VAppTemplate.Status != 8 {
+		return Task{}, fmt.Errorf("vApp Template shape is not ok (status: %d)", vappTemplate.VAppTemplate.Status)
+	}
+
+	// TODO - validate network *types.NetworkConnectionSection for mandatory parameters
```

Comment:
Are we sure on how to validate these parameters? If yes, meaning we won't put blockers on use cases we may not foresee, let's add validation and error messages now.

Reply:
Thanks, just one question to be safe: with the new validations it's still possible to add VM without a network connection still, correct?

Reply:
It accepted empty struct pointer `&types.NetworkConnectionSection{}`, but did not work if `nil` was passed for network. Added test for `nil` option as well. Last run and will push.

Reply:
Pushed